src/main/java/net/sf/jabref/gui/JabRefFrame.java

```diff
@@ -2337,4 +2282,104 @@ private int showSaveDialog(String filename) {
                 Localization.lang("Save before closing"), JOptionPane.YES_NO_CANCEL_OPTION,
                 JOptionPane.WARNING_MESSAGE, null, options, options[2]);
     }
+
+    private void closeTab(BasePanel panel) {
+        // TODO: this menu should be tab based not on the DragDropPopupPane
+        // empty tab without database
+        if (panel == null) {
+            return;
+        }
+
+        if (panel.isModified()) {
+            if(confirmClose(panel)) {
+                removeTab(panel);
+            }
+        } else {
+            removeTab(panel);
+        }
+    }
+
+    // Ask if the user really wants to close, if the base has not been saved
+    private boolean confirmClose(BasePanel panel) {
+        boolean close = false;
+        String filename;
+
+        if (panel.getDatabaseFile() != null) {
+            filename = panel.getDatabaseFile().getAbsolutePath();
+        } else {
+            filename = GUIGlobals.untitledTitle;
+        }
+
+        int answer = showSaveDialog(filename);
+        if (answer == JOptionPane.YES_OPTION) {
+            // The user wants to save.
+            try {
+                SaveDatabaseAction saveAction = new SaveDatabaseAction(panel);
+                saveAction.runCommand();
+                if (saveAction.isSuccess()) {
+                    close = true;
+                }
+            } catch (Throwable ex) {
+                // do not close
+            }
+
+        } else if(answer == JOptionPane.NO_OPTION) {
+            // discard changes
+            close = true;
+        }
+        return close;
+    }
+
+    private void removeTab(BasePanel panel) {
+        panel.cleanUp();
+        AutoSaveManager.deleteAutoSaveFile(panel);
+        tabbedPane.remove(panel);
+        if (tabbedPane.getTabCount() > 0) {
+            markActiveBasePanel();
+        }
+        setWindowTitle();
+        updateEnabledState(); // FIXME: Man, this is what I call a bug that this is not called.
```

Comment:
Maybe it is even fixed by calling this method and I added the FIXME because we must make sure that this is called every time we remove something, i don't know...